packages/models/src/lib/implementation/schemas.ts

```diff
@@ -167,26 +167,43 @@ export function scorableSchema<T extends ReturnType<typeof weightedRefSchema>>(
   duplicateCheckFn: (metrics: z.infer<T>[]) => false | string[],
   duplicateMessageFn: (metrics: z.infer<T>[]) => string,
 ) {
-  return z.object(
-    {
-      slug: slugSchema.describe('Human-readable unique ID, e.g. "performance"'),
-      refs: z
-        .array(refSchema)
-        .min(1)
-        // refs are unique
-        .refine(
-          refs => !duplicateCheckFn(refs),
-          refs => ({
-            message: duplicateMessageFn(refs),
-          }),
-        )
-        // categories weights are correct
-        .refine(hasNonZeroWeightedRef, () => ({
-          message:
-            'In a category there has to be at least one ref with weight > 0',
-        })),
-    },
-    { description },
+  return (
+    z
+      .object(
+        {
+          slug: slugSchema.describe(
+            'Human-readable unique ID, e.g. "performance"',
+          ),
+          refs: z
+            .array(refSchema)
+            .min(1)
+            // refs are unique
+            .refine(
+              refs => !duplicateCheckFn(refs),
+              refs => ({
+                message: duplicateMessageFn(refs),
+              }),
+            ),
+        },
+        { description },
+      )
+      // category weights are correct
+      .superRefine(({ slug, refs }, ctx) => {
+        if (refs.length === 0) {
+          ctx.addIssue({
+            code: z.ZodIssueCode.custom,
+            message: `In category ${slug}, there has to be at least one ref`,
+            path: ['refs'],
+          });
+        } else if (!hasNonZeroWeightedRef(refs)) {
+          const affectedRefs = refs.map(ref => ref.slug).join(', ');
+          ctx.addIssue({
+            code: z.ZodIssueCode.custom,
+            message: `In category ${slug}, there has to be at least one ref with weight > 0. Affected refs: ${affectedRefs}`,
```

Comment:
I would quote the slugs, make it more clear that they are dynamic values when reading the error message:

```suggestion
            message: `In category "${slug}", there has to be at least one ref`,
            path: ['refs'],
          });
        } else if (!hasNonZeroWeightedRef(refs)) {
          const affectedRefs = refs.map(ref => `"${ref.slug}"`).join(', ');
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            message: `In category "${slug}", there has to be at least one ref with weight > 0. Affected refs: ${affectedRefs}`,
```